### PR TITLE
Store execution result

### PR DESCRIPTION
**Pull request by shargon**
Close https://github.com/neo-project/neo/issues/2579

### CODE REVIEW

**Comment by shargon** [comment]
@erikzhang I think that this is an important change, do you prefer to change the type Block in ApplicationEngine to something like this?

```csharp
using Neo.VM;

namespace Neo.Network.P2P.Payloads
{
    public class PersistedBlock
    {
        /// <summary>
        /// The block to be persisted
        /// </summary>
        public readonly Block Block;

        /// <summary>
        /// The transaction's execution state list of the block.
        /// </summary>
        internal readonly VMState[] TransactionStates;

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="block">Block to be persisted</param>
        public PersistedBlock(Block block)
        {
            Block = block;
            TransactionStates = new VMState[block.Transactions.Length];
        }

        /// <summary>
        /// Index of the block
        /// </summary>
        public uint Index => Block.Index;
    }
}
```

**Comment by Liaojinghui** [comment]
I think it would be even better if we could have this result in the consensus period.

**Comment by shargon** [comment]
> I think it would be even better if we could have this result in the consensus period.

We can't

**Comment by Liaojinghui** [comment]
> We can't

that's sad

File: src/neo/SmartContract/Native/TransactionState.cs [hunk 1]
```diff
@@ -28,12 +33,13 @@ void IInteroperable.FromStackItem(StackItem stackItem)
             BlockIndex = (uint)@struct[0].GetInteger();
             _rawTransaction = @struct[1];
             Transaction = _rawTransaction.GetSpan().AsSerializable<Transaction>();
+            State = @struct.Count == 2 ? VMState.NONE : (VMState)(byte)@struct[2].GetInteger();
```
[the remark this answers is not in the document]
Comment (erikzhang) [reply src/neo/SmartContract/Native/TransactionState.cs:46]: We can resync.

**Comment by erikzhang** [comment]
@shargon Please check my last changes.

File: src/neo/SmartContract/Native/TransactionState.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (shargon) [reply src/neo/SmartContract/Native/TransactionState.cs:46]:
```suggestion
            State = (VMState)(byte)@struct[2].GetInteger();
```

File: src/neo/Ledger/Blockchain.cs [hunk 2]
```diff
@@ -420,9 +423,16 @@ private void Persist(Block block)
...
+                    transactionStates[i] = new TransactionState
+                    {
+                        BlockIndex = block.Index,
+                        Transaction = tx,
+                        State = state
+                    };
                 }
                 using (ApplicationEngine engine = ApplicationEngine.Create(TriggerType.PostPersist, null, snapshot, block, system.Settings, 0))
                 {
+                    engine.SetState(transactionStates);
```
Comment (shargon) [inline src/neo/Ledger/Blockchain.cs:435]: @erikzhang Maybe we can remove block from `Create` and use a state?

Comment (erikzhang) [reply src/neo/Ledger/Blockchain.cs:435]: But many services rely on block.

**Review by erikzhang — DISMISSED** [review]

**Comment by shargon** [comment]
@superboyiii could you test it?

**Review by erikzhang — APPROVED** [review]

File: src/neo/Ledger/Blockchain.cs [hunk 3]
```diff
@@ -395,21 +395,25 @@ private void Persist(Block block)
...
                     if (engine.Execute() != VMState.HALT) throw new InvalidOperationException();
                     ApplicationExecuted application_executed = new(engine);
                     Context.System.EventStream.Publish(application_executed);
                     all_application_executed.Add(application_executed);
+                    transactionStates = engine.GetState<TransactionState[]>();
                 }
                 DataCache clonedSnapshot = snapshot.CreateSnapshot();
                 // Warning: Do not write into variable snapshot directly. Write into variable clonedSnapshot and commit instead.
-                foreach (Transaction tx in block.Transactions)
+                foreach (TransactionState transactionState in transactionStates)
```
Comment (shargon) [inline src/neo/Ledger/Blockchain.cs:410]: Add an assert for check that TransactionState always is equal to blocks.Transactions?

Comment (erikzhang) [reply src/neo/Ledger/Blockchain.cs:410]: I don't think it is necessary.

**Comment by superboyiii** [comment]
Because of https://github.com/neo-project/neo/pull/2613, we have to fix compiler first.
5949a6ba53605d2421d0a89fa94d38c
@erikzhang @shargon

**Comment by erikzhang** [comment]
> we have to fix compiler first.

You can test it without the compiler.

**Comment by superboyiii** [comment]
**Test PASS**
I made a contract and try to debug it. Yes, the state can be showed correctly.

**Review by superboyiii — APPROVED** [review]

**Pull request merged**